Content.Server/Explosion/EntitySystems/ClusterGrenadeSystem.cs

```diff
@@ -58,59 +60,86 @@ private void OnClugUsing(Entity<ClusterGrenadeComponent> clug, ref InteractUsing
-
-                grenade.SpawnTimer(delay, () =>
+                while (TryGetGrenade(uid, clug, out var grenade))
                 {
-                    if ((!EntityManager.EntityExists(grenade) ? EntityLifeStage.Deleted : MetaData(grenade).EntityLifeStage) >= EntityLifeStage.Deleted)
-                        return;
-
-                    _trigger.Trigger(grenade, args.User);
-                });
+                    // var distance = random.NextFloat() * _throwDistance;
+                    var angleMin = segmentAngle * thrownCount;
+                    var angleMax = segmentAngle * (thrownCount + 1);
+                    var angle = Angle.FromDegrees(_random.Next(angleMin, angleMax));
+                    if (clug.RandomAngle)
+                        angle = _random.NextAngle();
+                    thrownCount++;
+
+                    switch (clug.GrenadeType)
+                    {
+                        case "shoot":
+                            ShootProjectile(grenade, angle, clug, uid);
+                            break;
+                        case "throw":
+                            ThrowGrenade(grenade, angle, clug);
+                            break;
+                    }
+
+                    // give an active timer trigger to the contained grenades when they get launched
+                    if (clug.TriggerBomblets)
+                    {
+                        bombletDelay += _random.NextFloat(clug.BombletDelayMin, clug.BombletDelayMax);
+                        var bomblet = EnsureComp<ActiveTimerTriggerComponent>(grenade);
+                        bomblet.TimeRemaining = (clug.MinimumDelay + bombletDelay);
+                        var ev = new ActiveTimerTriggerEvent(grenade, uid);
+                        RaiseLocalEvent(uid, ref ev);
+                    }
+                }
+                // delete the empty shell of the clusterbomb
+                EntityManager.DeleteEntity(uid);
             }
+        }
+    }
 
-            EntityManager.DeleteEntity(uid);
-        });
+    private void ShootProjectile(EntityUid grenade, Angle angle, ClusterGrenadeComponent clug, EntityUid clugUid)
+    {
+        if (clug.RandomSpread)
+            _gun.ShootProjectile(grenade, _random.NextVector2().Normalized(), Vector2.One.Normalized(), clugUid);
+        else _gun.ShootProjectile(grenade, angle.ToVec().Normalized(), Vector2.One.Normalized(), clugUid);
 
-        args.Handled = true;
     }
 
-    private bool TryGetGrenade(Entity<ClusterGrenadeComponent> ent, out EntityUid grenade)
+    private void ThrowGrenade(EntityUid grenade, Angle angle, ClusterGrenadeComponent clug)
+    {
+        if (clug.RandomSpread)
+            _throwingSystem.TryThrow(grenade, angle.ToVec().Normalized() * _random.NextFloat(clug.MinSpreadDistance, clug.MaxSpreadDistance), clug.Velocity);
+        else _throwingSystem.TryThrow(grenade, angle.ToVec().Normalized() * clug.Distance, clug.Velocity);
+    }
```

Comment:
the formatting on both of these is kinda weird. you could just do an angle variable with a ternary operator and it'd look a lot cleaner.

Reply:
Changed the formatting